src/Raven.Client/Documents/Session/AsyncDocumentSession.cs

```diff
@@ -56,15 +59,44 @@ public async Task RefreshAsync<T>(T entity, CancellationToken token = default(Ca
         {
             using (AsyncTaskHolder())
             {
-                DocumentInfo documentInfo;
-                if (DocumentsByEntity.TryGetValue(entity, out documentInfo) == false)
+                if (DocumentsByEntity.TryGetValue(entity, out var documentInfo) == false)
                     throw new InvalidOperationException("Cannot refresh a transient instance");
                 IncrementRequestCount();
 
                 var command = new GetDocumentsCommand(new[] { documentInfo.Id }, includes: null, metadataOnly: false);
                 await RequestExecutor.ExecuteAsync(command, Context, _sessionInfo, token).ConfigureAwait(false);
 
-                RefreshInternal(entity, command, documentInfo);
+                var commandResult = (BlittableJsonReaderObject)command.Result.Results[0];
+                RefreshInternal(entity, commandResult, documentInfo);
+            }
+        }
+
+        public async Task RefreshAsync<T>(IEnumerable<T> entities, CancellationToken token = default)
+        {
+            using (AsyncTaskHolder())
+            {
+                var idsEntitiesPairs = new Dictionary<string, (object Entity, DocumentInfo Info)>();
+
+                foreach (var entity in entities)
+                {
+                    if (DocumentsByEntity.TryGetValue(entity, out var docInfo) == false)
+                        throw new InvalidOperationException("Cannot refresh a transient instance");
+                    idsEntitiesPairs[docInfo.Id] = (entity, docInfo);
+                }
+                IncrementRequestCount();
+
+                var command = new GetDocumentsCommand(idsEntitiesPairs.Keys.ToArray(), includes: null, metadataOnly: false);
+                await RequestExecutor.ExecuteAsync(command, Context, sessionInfo: _sessionInfo, token).ConfigureAwait(false);
+
+                var resultsCollection = command.Result.Results;
```

Comment:
What happens if the document that is being refreshed was deleted in the meantime?

Reply:
a NullReferenceException will be thrown few lines down when trying to access its elements.
handled it by throwing a custom exception and added some scenarios to test this particular situation.